### PR TITLE
Uncancel tasks cancelled using CancelScope.cancel()

### PR DESCRIPTION
**Pull request by nikicat**
This patch is needed for correct task canceling when using asyncio backend on Python >= 3.11.
Related discussion https://github.com/encode/httpx/discussions/2506

### CODE REVIEW

**Comment by graingert** [comment]
uncancel needs to be called each time anyio calls cancel() not just the one time

File: tests/test_taskgroups.py [hunk 1]
```diff
@@ -1092,3 +1092,15 @@ async def starter_task() -> None:
...
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 11),
+    reason="Task uncancelling is only supported on Python 3.11",
+)
+@pytest.mark.parametrize("anyio_backend", ["asyncio"])
+async def test_uncancel_after_cancel() -> None:
+    with CancelScope() as scope:
+        scope.cancel()
+        await anyio.sleep(0)
```
Comment (graingert) [inline tests/test_taskgroups.py:1105]:
```suggestion
        with pytest.raises(asyncio.CancelledError):
            await anyio.sleep(0)
        with pytest.raises(asyncio.CancelledError):
            await anyio.sleep(0)
```

File: src/anyio/_backends/_asyncio.py [hunk 2]
```diff
@@ -270,13 +270,20 @@ def __exit__(
                 exc_val.exceptions if isinstance(exc_val, ExceptionGroup) else [exc_val]
             )
             if all(isinstance(exc, CancelledError) for exc in exceptions):
+
+                def uncancel() -> None:
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:274]: can you copy the existing pattern for version compat:

https://github.com/agronholm/anyio/blob/e2ff31573aa14df5abfb96b3c56559886310fe36/src/anyio/_backends/_asyncio.py#L86-L94
```

**Comment by nikicat** [comment]
> uncancel needs to be called each time anyio calls cancel() not just the one time

I thought `.uncancel()` should be only called in `__exit__` and only when it suppressses `CancelledError` exception.

**Comment by nikicat** [comment]
> uncancel needs to be called each time anyio calls cancel() not just the one time

I thought `.uncancel()` should be only called in `__exit__` and only when it suppressses `CancelledError` exception.

**Comment by graingert** [comment]
yeah so you need something like:

```python
for i in range(cancel_calls):
    task.uncancel()
```

File: src/anyio/_backends/_asyncio.py [hunk 3]
```diff
@@ -270,13 +270,20 @@ def __exit__(
...
+                    if hasattr(self._host_task, "uncancel"):
+                        self._host_task.uncancel()  # type: ignore[union-attr]
+
                 if self._timeout_expired:
+                    uncancel()
                     return True
                 elif not self._cancel_called:
                     # Task was cancelled natively
                     return None
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:283]: native cancellation on 3.11 can now be detected by comparing the number of times anyio called .cance() and the state of `task.cancelling()`

**Comment by agronholm** [comment]
A fix for uncancel() was included in my local WIP branch before I switched to other projects. Now that I've resumed working on AnyIO, I have to see which patches make the most sense.

**Comment by nikicat** [comment]
I have updated this PR according to your suggestions. The only aspect I am uncertain about is allowing multiple context entries for a `CancelScope`. In the current implementation `CancelScope` remains in a canceled state after exiting.

**Comment by agronholm** [comment]
I'm not sure that you should hurry with this. I did have code written to tackle this issue, so it's possible I might end up rejecting this PR. I will, however, take a look at it and see if I can pick up any good ideas.

File: src/anyio/_backends/_asyncio.py (hunk 3 above)
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:278]: I've added detection of native cancellation, and a test to that end, on OP's behalf.

**Comment by agronholm** [comment]
@nikicat I took the liberty of making some improvements to this PR. Would you like to do a sanity check on my changes so we can get this merged?

**Comment by nikicat** [comment]
@agronholm looks good, LGTM! Thank you!

**Review by agronholm — APPROVED** [review]

**Comment by agronholm** [comment]
Thanks!

**Pull request merged**